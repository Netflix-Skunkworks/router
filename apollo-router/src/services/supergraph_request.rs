pub(crate) mod service;

use super::supergraph::{Request, Response};

pub type BoxService = tower::util::BoxService<Request, Request, Response>;
