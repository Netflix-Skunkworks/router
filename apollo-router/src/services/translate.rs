use std::task::Poll;
use futures::future::BoxFuture;
use tower::Service;

use super::supergraph::Request;
use super::supergraph::Response;

pub type BoxService = tower::util::BoxService<Request, Request, Response>;
pub type BoxCloneService = tower::util::BoxCloneService<Request, Request, Response>;
pub type ServiceResult = Result<Request, Response>;

pub struct TranslateService;

impl Service<Request> for TranslateService {
    type Response = Request;
    type Error = Response;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: Request) -> Self::Future {
        Box::pin(async { Ok(req) })
    }
}
