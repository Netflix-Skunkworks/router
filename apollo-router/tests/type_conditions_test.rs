//!
//! Please ensure that any tests added to this file use the tokio multi-threaded test executor.
//!

use apollo_router::plugin::test::MockSubgraph;
use apollo_router::services::supergraph;
use apollo_router::{MockedSubgraphs, TestHarness};
use serde_json::json;
use tower::ServiceExt;

mod integration;


#[tokio::test(flavor = "multi_thread")]
async fn test_type_conditions_enabled() {
    let harness = setup(json! {{
        "experimental_type_conditioned_fetching": true
    }});
    let supergraph_service = harness.build_supergraph().await.unwrap();
    let request = supergraph::Request::fake_builder()
        .query(QUERY.to_string())
        .build()
        .expect("expecting valid request")
        .try_into()
        .unwrap();

    let response = supergraph_service
        .oneshot(request)
        .await
        .unwrap()
        .next_response()
        .await
        .unwrap();

    insta::assert_json_snapshot!(response);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_type_conditions_disabled() {
    let harness = setup(json! {{
        "experimental_type_conditioned_fetching": false
    }});
    let supergraph_service = harness.build_supergraph().await.unwrap();
    let request = supergraph::Request::fake_builder()
        .query(QUERY.to_string())
        .build()
        .expect("expecting valid request")
        .try_into()
        .unwrap();

    let response = supergraph_service
        .oneshot(request)
        .await
        .unwrap()
        .next_response()
        .await
        .unwrap();

    insta::assert_json_snapshot!(response);
}

fn setup(configuration: serde_json::Value) -> TestHarness<'static> {
    let search_service =  MockSubgraph::builder().with_json(json!{{
        "query":"query Search__searchSubgraph__0{search{__typename ...on MovieResult{sections{__typename ...on EntityCollectionSection{__typename id}...on GallerySection{__typename id}}id}...on ArticleResult{id sections{__typename ...on GallerySection{__typename id}...on EntityCollectionSection{__typename id}}}}}",
        "operationName":"Search__searchSubgraph__0"
    }},
json!{{
        "data": {
            "search":[
                {
                    "__typename":"ArticleResult",
                    "id":"ff70d1f5-d1ac-46dd-8ed1-5f2d81ff2db0",
                    "sections":[
                        {
                            "__typename":"EntityCollectionSection",
                            "id":"a7487f33-bd37-48a6-a843-c0cda86f5049"
                        },
                        {
                            "__typename":"EntityCollectionSection",
                            "id":"cdb43f1d-df2d-4293-a328-8d38a0cdd742"
                        }
                    ]
                },
                {
                    "__typename":"ArticleResult",
                    "id":"5092bbea-8bc3-4c4f-a9eb-003604ed9add",
                    "sections":[
                        {
                            "__typename":"GallerySection",
                            "id":"798e75ae-9378-41de-a014-af9f9a5e99eb"
                        },
                        {
                            "__typename":"GallerySection",
                            "id":"f756501a-7377-4081-861b-0097cbfb7f41"
                        }
                    ]
                }
            ]
        }
    }}).build();

    let artwork_service = MockSubgraph::builder().with_json(json!{{
        "query":"query Search__artworkSubgraph__1($representations:[_Any!]!){_entities(representations:$representations){...on EntityCollectionSection{artwork title}...on GallerySection{artwork}}}","operationName":"Search__artworkSubgraph__1",
        "variables":{
            "representations":[
                {
                    "__typename":"EntityCollectionSection",
                    "id":"a7487f33-bd37-48a6-a843-c0cda86f5049"
                },
                {
                    "__typename":"EntityCollectionSection",
                    "id":"cdb43f1d-df2d-4293-a328-8d38a0cdd742"
                },
                {
                    "__typename":"GallerySection",
                    "id":"798e75ae-9378-41de-a014-af9f9a5e99eb"
                },
                {
                    "__typename":"GallerySection",
                    "id":"f756501a-7377-4081-861b-0097cbfb7f41"
                }
            ]
        }
    }},
json!{{
        "data":{
            "_entities":[
                {
                    "artwork":"Hello World",
                    "title":"Hello World"
                },
                {
                    "artwork":"Hello World",
                    "title":"Hello World"
                }
            ]
        }
    }}).build();

    let mut mocks = MockedSubgraphs::default();
    mocks.insert("searchSubgraph", search_service);
    mocks.insert("artworkSubgraph", artwork_service);

    let schema = include_str!("fixtures/type_conditions.graphql");
    TestHarness::builder()
        .try_log_level("info")
        .configuration_json(configuration)
        .unwrap()
        .schema(schema)
        .extra_plugin(mocks)
}

static QUERY: &str = "
query Search {
    search {
      ... on MovieResult {
        sections {
          ... on EntityCollectionSection {
            artwork
            id
            title
          }
          ... on GallerySection {
            artwork
            id
          }
        }
        id
      }
      ... on ArticleResult {
        id
        sections {
          ... on GallerySection {
            artwork
          }
          ... on EntityCollectionSection {
            artwork
          }
        }
      }
    }
}";
