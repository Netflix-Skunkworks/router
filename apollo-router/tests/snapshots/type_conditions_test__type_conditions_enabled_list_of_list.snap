---
source: apollo-router/tests/type_conditions_test.rs
expression: response
---
{
  "data": {
    "searchListOfList": [
      [
        null,
        null,
        null,
        null
      ],
      [
        null,
        {}
      ],
      [
        null,
        null
      ]
    ]
  },
  "extensions": {
    "apolloQueryPlan": {
      "object": {
        "kind": "QueryPlan",
        "node": {
          "kind": "Sequence",
          "nodes": [
            {
              "kind": "Fetch",
              "serviceName": "searchSubgraph",
              "variableUsages": [],
              "operation": "query Search__searchSubgraph__0{searchListOfList{__typename ...on MovieResult{sections{__typename ...on EntityCollectionSection{__typename id}...on GallerySection{__typename id}}id}...on ArticleResult{id sections{__typename ...on GallerySection{__typename id}...on EntityCollectionSection{__typename id}}}}}",
              "operationName": "Search__searchSubgraph__0",
              "operationKind": "query",
              "id": null,
              "inputRewrites": null,
              "outputRewrites": null,
              "schemaAwareHash": "2e77569c51009a58e162f112a331bb32b4ee615de2911455d9b0e533489859b2",
              "authorization": {
                "is_authenticated": false,
                "scopes": [],
                "policies": []
              }
            },
            {
              "kind": "Parallel",
              "nodes": [
                {
                  "kind": "Flatten",
                  "path": [
                    "",
                    "searchListOfList",
                    "@",
                    "@|[MovieResult]",
                    "sections",
                    "@"
                  ],
                  "node": {
                    "kind": "Fetch",
                    "serviceName": "artworkSubgraph",
                    "requires": [
                      {
                        "kind": "InlineFragment",
                        "typeCondition": "EntityCollectionSection",
                        "selections": [
                          {
                            "kind": "Field",
                            "name": "__typename"
                          },
                          {
                            "kind": "Field",
                            "name": "id"
                          }
                        ]
                      },
                      {
                        "kind": "InlineFragment",
                        "typeCondition": "GallerySection",
                        "selections": [
                          {
                            "kind": "Field",
                            "name": "__typename"
                          },
                          {
                            "kind": "Field",
                            "name": "id"
                          }
                        ]
                      }
                    ],
                    "variableUsages": [
                      "movieResultParam"
                    ],
                    "operation": "query Search__artworkSubgraph__1($representations:[_Any!]!$movieResultParam:String){_entities(representations:$representations){...on EntityCollectionSection{title artwork(params:$movieResultParam)}...on GallerySection{artwork(params:$movieResultParam)}}}",
                    "operationName": "Search__artworkSubgraph__1",
                    "operationKind": "query",
                    "id": null,
                    "inputRewrites": null,
                    "outputRewrites": null,
                    "schemaAwareHash": "68cc2785b5f8b0fb994ba1d29b86bfebcbdeb005eb62d078abd879e0f3404b11",
                    "authorization": {
                      "is_authenticated": false,
                      "scopes": [],
                      "policies": []
                    }
                  }
                },
                {
                  "kind": "Flatten",
                  "path": [
                    "",
                    "searchListOfList",
                    "@",
                    "@|[ArticleResult]",
                    "sections",
                    "@"
                  ],
                  "node": {
                    "kind": "Fetch",
                    "serviceName": "artworkSubgraph",
                    "requires": [
                      {
                        "kind": "InlineFragment",
                        "typeCondition": "GallerySection",
                        "selections": [
                          {
                            "kind": "Field",
                            "name": "__typename"
                          },
                          {
                            "kind": "Field",
                            "name": "id"
                          }
                        ]
                      },
                      {
                        "kind": "InlineFragment",
                        "typeCondition": "EntityCollectionSection",
                        "selections": [
                          {
                            "kind": "Field",
                            "name": "__typename"
                          },
                          {
                            "kind": "Field",
                            "name": "id"
                          }
                        ]
                      }
                    ],
                    "variableUsages": [
                      "articleResultParam"
                    ],
                    "operation": "query Search__artworkSubgraph__2($representations:[_Any!]!$articleResultParam:String){_entities(representations:$representations){...on GallerySection{artwork(params:$articleResultParam)}...on EntityCollectionSection{artwork(params:$articleResultParam)title}}}",
                    "operationName": "Search__artworkSubgraph__2",
                    "operationKind": "query",
                    "id": null,
                    "inputRewrites": null,
                    "outputRewrites": null,
                    "schemaAwareHash": "1f47d603f0bc93220aac7806a8d203a5595bcb9e98ceacc3c4a9ba3951e16a5b",
                    "authorization": {
                      "is_authenticated": false,
                      "scopes": [],
                      "policies": []
                    }
                  }
                }
              ]
            }
          ]
        }
      },
      "text": "QueryPlan {\n  Sequence {\n    Fetch(service: \"searchSubgraph\") {\n      {\n        searchListOfList {\n          __typename\n          ... on MovieResult {\n            sections {\n              __typename\n              ... on EntityCollectionSection {\n                __typename\n                id\n              }\n              ... on GallerySection {\n                __typename\n                id\n              }\n            }\n            id\n          }\n          ... on ArticleResult {\n            id\n            sections {\n              __typename\n              ... on GallerySection {\n                __typename\n                id\n              }\n              ... on EntityCollectionSection {\n                __typename\n                id\n              }\n            }\n          }\n        }\n      }\n    },\n    Parallel {\n      Flatten(path: \".searchListOfList.@.@|[MovieResult].sections.@\") {\n        Fetch(service: \"artworkSubgraph\") {\n          {\n            ... on EntityCollectionSection {\n              __typename\n              id\n            }\n            ... on GallerySection {\n              __typename\n              id\n            }\n          } =>\n          {\n            ... on EntityCollectionSection {\n              title\n              artwork(params: $movieResultParam)\n            }\n            ... on GallerySection {\n              artwork(params: $movieResultParam)\n            }\n          }\n        },\n      },\n      Flatten(path: \".searchListOfList.@.@|[ArticleResult].sections.@\") {\n        Fetch(service: \"artworkSubgraph\") {\n          {\n            ... on GallerySection {\n              __typename\n              id\n            }\n            ... on EntityCollectionSection {\n              __typename\n              id\n            }\n          } =>\n          {\n            ... on GallerySection {\n              artwork(params: $articleResultParam)\n            }\n            ... on EntityCollectionSection {\n              artwork(params: $articleResultParam)\n              title\n            }\n          }\n        },\n      },\n    },\n  },\n}"
    }
  }
}
